x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -170,6 +183,76 @@ private <T> void preAnalyzeIndices(LogicalPlan parsed, ActionListener<IndexResol
         }
     }
 
+    static Set<String> fieldNames(LogicalPlan parsed) {
+        if (false == parsed.anyMatch(plan -> plan instanceof Aggregate || plan instanceof Project)) {
+            // no explicit columns selection, for example "from employees"
+            return IndexResolver.ALL_FIELDS;
+        }
+
+        Holder<Boolean> projectAll = new Holder<>(false);
+        parsed.forEachExpressionDown(UnresolvedStar.class, us -> {// explicit "*" fields selection
+            if (projectAll.get()) {
+                return;
+            }
+            projectAll.set(true);
+        });
+        if (projectAll.get()) {
+            return IndexResolver.ALL_FIELDS;
+        }
+
+        AttributeSet references = new AttributeSet();
+        // "keep" attributes are special whenever a wildcard is used in their name
+        // ie "from test | eval lang = languages + 1 | keep *l" should consider both "languages" and "*l" as valid fields to ask for
+        AttributeSet keepCommandReferences = new AttributeSet();
+
+        parsed.forEachDown(p -> {// go over each plan top-down
+            if (p instanceof RegexExtract re) { // for Grok and Dissect
+                AttributeSet dissectRefs = p.references();
+                // don't add to the list of fields the extracted ones (they are not real fields in mappings)
+                dissectRefs.subtract(new AttributeSet(re.extractedFields()));
+                references.addAll(dissectRefs);
+                // also remove other down-the-tree references to the extracted fields
+                for (Attribute extracted : re.extractedFields()) {
+                    references.removeIf(attr -> isAttributeEqualTo(attr, extracted.qualifiedName(), false));
+                }
+            } else {
+                references.addAll(p.references());
+                if (p instanceof Keep) {
```

Comment:
What about drop - should its argument be removed from the list of fields requested?

Reply:
I didn't, on purpose, look at removing things too aggressively from the list of fields with the exception of this [particular use case](https://github.com/elastic/elasticsearch/pull/98987/files#diff-40060e2ec9003953a228c4a03bdc80a301949b0d4e3dccc1978798e33a992a73R238-R240). Also, I considered the fields in a `drop` command to be "valid", meaning the ResolveRefs rule in the Analyzer would want to know if `drop first_name` command refers to a valid `first_name` field. There are probably use cases where `drop` refers to an alias which should be safe to remove. For example `from test | eval alias_to_first_name = first_name | drop alias_to_first_name`  is a scenario where `alias_to_first_name` shouldn't appear in the list of fields. And I don't think it does appear in the list of fields in the PR's current form (since the code catches `alias_to_first_name` and removes it).